File: shell/com/apkharden/shell/AntiDebug.java
package com.apkharden.shell;

import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.os.Debug;
import java.io.BufferedReader;
import java.io.FileReader;

final class AntiDebug {
    /** @return true if a debugger / tracer is detected. */
    static boolean isDetected(Context ctx) {
        if (Debug.isDebuggerConnected() || Debug.waitingForDebugger()) return true;
        if ((ctx.getApplicationInfo().flags & ApplicationInfo.FLAG_DEBUGGABLE) != 0) return true;
        BufferedReader r = null;
        try {
            r = new BufferedReader(new FileReader("/proc/self/status"));
            String line;
            while ((line = r.readLine()) != null) {
                if (line.startsWith("TracerPid:")) {
                    String v = line.substring("TracerPid:".length()).trim();
                    if (!"0".equals(v)) return true;
                    break;
                }
            }
        } catch (Exception ignored) {
        } finally {
            if (r != null) try { r.close(); } catch (Exception ignored) {}
        }
        return false;
    }

    private AntiDebug() {}
}

File: harden-runtime/src/main/kotlin/com/apkharden/runtime/AntiDebug.kt
package com.apkharden.runtime

import android.content.Context
import android.content.pm.ApplicationInfo
import android.os.Debug
import java.io.File

object AntiDebug {
    fun detect(context: Context): HardenFailure? = when {
        context.applicationInfo.flags and ApplicationInfo.FLAG_DEBUGGABLE != 0 ->
            HardenFailure.DEBUGGABLE
        Debug.isDebuggerConnected() || Debug.waitingForDebugger() ->
            HardenFailure.DEBUGGER_CONNECTED
        readTracerPid() -> HardenFailure.TRACER_DETECTED
        else -> null
    }

    internal fun hasTracer(lines: Sequence<String>): Boolean = lines
        .firstOrNull { it.startsWith("TracerPid:") }
        ?.substringAfter(':')
        ?.trim()
        ?.toIntOrNull()
        ?.let { it > 0 }
        ?: false

    private fun readTracerPid(): Boolean = runCatching {
        File("/proc/self/status").useLines { lines -> hasTracer(lines) }
    }.getOrDefault(false)
}
